include/deal.II/lac/solver_cg.h

```diff
@@ -1042,6 +1074,32 @@ namespace internal
                 v += length;
               }
           }
+        else if (iteration_index % 2 == 0 && beta == Number())
+          {
+            for (unsigned int j = start_range; j < end_range; j += grain_size)
+              {
+                const unsigned int length = std::min(grain_size, end_range - j);
+                DEAL_II_OPENMP_SIMD_PRAGMA
+                for (unsigned int i = 0; i < length; ++i)
+                  r[i] -= this->alpha * v[i];
+                DEAL_II_OPENMP_SIMD_PRAGMA
+                for (unsigned int i = 0; i < length; ++i)
+                  x[i] += this->alpha * p[i];
+                this->preconditioner.apply_to_subrange(j,
+                                                       j + length,
+                                                       r,
+                                                       prec_r.data());
+                DEAL_II_OPENMP_SIMD_PRAGMA
+                for (unsigned int i = 0; i < length; ++i)
+                  {
+                    p[i] = this->beta * p[i] + prec_r[i];
+                    v[i] = Number();
+                  }
+                p += length;
+                r += length;
+                v += length;
+              }
+          }
```

Comment:
Isn't this the exact same code as for `iteration_index % 2 == 0 && beta != Number()` or did you mean to exploit `beta==0` in line 1095?

Reply:
It is almost the same, we add the line to compute `x` (line 1087). I add the comment from above and restructure the code slightly.